pkg/agent/http.go

```diff
@@ -59,16 +62,57 @@ func accessControl(agt *agent, proxyHandler http.Handler) http.Handler {
 
 	router.Use(func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			//user header
+			rancherUser := r.Header.Get(rancherUserHeaderKey)
+			rancherGroup := r.Header[rancherGroupHeaderKey]
+
+			//sa header
 			accessToken := strings.TrimPrefix(r.Header.Get(authorizationHeaderKey), "Bearer ")
-			if len(accessToken) == 0 {
+
+			if rancherUser == "" && len(rancherGroup) == 0 && len(accessToken) == 0 {
 				http.Error(w, "unauthorized", http.StatusUnauthorized)
 				return
 			}
 
-			// direct proxy
-			if agt.cfg.myToken == accessToken {
-				proxyHandler.ServeHTTP(w, r)
-				return
+			var namespaceSet data.Set
+			if rancherUser != "" || len(rancherGroup) != 0 {
+				log.Debugf("%s - %s - access by userID", r.Method, r.URL.Path)
+				info := &user.DefaultInfo{
+					Name:   rancherUser,
+					UID:    rancherUser,
+					Groups: rancherGroup,
+				}
+
+				if agt.nodes.CanList(info) {
```

Comment:
Code duplicated in Line 110-115. You can put them out of the if-else block.

Reply:
updated